api/src/test/java/com/epam/pipeline/manager/pipeline/FolderTemplateManagerTest.java

```diff
@@ -75,98 +67,130 @@
     private static final String CHILD_TEMPLATE_FOLDER_NAME_1 = "test-1";
     private static final String DATASTORAGE_NAME_1 = "ds-1";
 
+    private static final Long FOLDER_ID = 1L;
+    private static final Long FOLDER_CHILD_ID = 2L;
+
     @Autowired
     private FolderTemplateManager folderTemplateManager;
     @Autowired
-    private FolderManager folderManager;
-    @Autowired
-    private MetadataManager metadataManager;
-    @Autowired
-    private DataStorageManager dataStorageManager;
-    @SpyBean
-    private S3StorageProvider storageProviderManager;
+    private MetadataManager mockMetadataManager;
     @Autowired
-    private AclTestDao aclTestDao;
+    private DataStorageManager mockDataStorageManager;
     @Autowired
-    private GrantPermissionManager permissionManager;
+    private GrantPermissionManager spyGrantPermissionManager;
     @Autowired
-    private CloudRegionDao cloudRegionDao;
+    private FolderCrudManager mockFolderCrudManager;
 
-    private AwsRegion awsRegion;
+    private Map<String, PipeConfValue> metadata;
+    private PermissionVO permissionVO;
+    private FolderTemplate folderTemplate;
 
+    @Test
+    public void createFolderFromTemplateTest() throws IOException {
+        initializeParameters();
+        preConditions();
 
-    @Before
-    public void setUp() {
-        doReturn(new MockS3Helper()).when(storageProviderManager).getS3Helper(any(S3bucketDataStorage.class));
+        Folder folder = new Folder();
+        folder.setName(TEST_NAME_2);
+        folder.setId(FOLDER_ID);
+        folderTemplateManager.createFolderFromTemplate(folder, folderTemplate);
 
-        awsRegion = new AwsRegion();
-        awsRegion.setName("US");
-        awsRegion.setRegionCode("us-east-1");
-        awsRegion.setDefault(true);
-        cloudRegionDao.create(awsRegion);
+        Assert.assertEquals(TEMPLATE_FOLDER_NAME, folder.getName());
 
-        AclTestDao.AclSid testUserSid = new AclTestDao.AclSid(true, TEST_USER);
-        aclTestDao.createAclSid(testUserSid);
-        AclTestDao.AclSid testRole = new AclTestDao.AclSid(false, TEST_ROLE);
-        aclTestDao.createAclSid(testRole);
+        Assert.assertTrue(folderTemplate.getDatastorages()
+                .stream()
+                .allMatch(storage -> FOLDER_ID.equals(storage.getParentFolderId())));
+        Assert.assertTrue(folderTemplate.getChildren()
+                .stream()
+                .flatMap(childTemplate -> childTemplate.getDatastorages().stream())
+                .allMatch(storage -> FOLDER_CHILD_ID.equals(storage.getParentFolderId())));
+        Assert.assertEquals(metadata, folderTemplate.getMetadata());
+        Assert.assertEquals(1, folderTemplate.getPermissions().size());
```

Comment:
The idea behind `FolderTemplateManager.createFolderFromTemplate` is to create folder and storages structure according to a folder template. These asserts test that the given folder template is modified by the original method. You see, this modification is just a side effect. But the core functionality that should be tested is which folders or storages were created within which folders.

So I would suggest replacing such asserts with more precise verification calls.

Reply:
Fixed